pymeasure/experiment/results.py

```diff
@@ -129,16 +131,54 @@ def __init__(self, columns, delimiter=','):
         """
         super().__init__()
         self.columns = columns
+        self.units = self._parse_columns(columns)
         self.delimiter = delimiter
 
+    @staticmethod
+    def _parse_columns(columns):
+        """Parse the columns to get units in parenthesis."""
+        units_pattern = r"\((?P<units>[\w/\(\)\*\t]+)\)"
+        units = {}
+        for column in columns:
+            try:
+                match = re.search(units_pattern, column)
+            except TypeError:
+                match = None
+
+            if match:
+                m = match.groupdict()
+                if 'units' in m:
+                    units[column] = pint.Unit(m['units'])
+        return units
+
     def format(self, record):
         """Formats a record as csv.
 
         :param record: record to format.
         :type record: dict
         :return: a string
         """
-        return self.delimiter.join(f'{record[x]}' for x in self.columns)
+        line = []
+        for x in self.columns:
+            value = record.get(x, float("nan"))
+            units = self.units.get(x, u.dimensionless)
+            if isinstance(value, str):
+                try:
+                    value = pint.Quantity(value)
+                except pint.UndefinedUnitError:
+                    pass  # Keep it as a string
+            if isinstance(value, pint.Quantity):
+                try:
+                    line.append(f"{value.m_as(units)}")
+                except pint.DimensionalityError:
+                    line.append("nan")
+            elif units == u.dimensionless:
+                # Dimensionless quantities can be given as numbers.
```

Comment:
Logic alright here? A dimensionless quantity would match in the previous clause (as it is a pint.Quantity) already. I suspect this clause can be dropped?